File: Form1.cs
using System;
using System.Windows.Forms;

namespace Expresiones_Lambda_WinFormsApp1_cs
{
    public partial class Form1 : Form
    {
        Form1()
        {
            InitializeComponent();
            this.Load += Form1_Load;
            richTextBoxCodigo = new RichTextBox();
            buttonAbrir = new Button();
            buttonUndo = new Button();
            menuUndo = new ToolStripMenuItem();
        }

        [STAThread]
        static void Main()
        {
            Application.SetHighDpiMode(HighDpiMode.SystemAware);
            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);
            Application.Run(new Form1());
        }

        private Button buttonAbrir;
        //private RichTextBox richTextBoxCodigo;
        private Button buttonUndo;
        private ToolStripMenuItem menuUndo;

        // En C# no permite declararlo fuera del cuerpo de un método
        // si no, da error de que el control no está definido (debe ser static)
        // A field initializer cannot reference the non-static field, method or property 'Form1.richTextBoxCodigo'
        // por tanto: definiendo el richtextbox como static ya funciona

        private static RichTextBox richTextBoxCodigo;

        // No hay porqué definir los tipos de los parámetros
        private EventHandler lambdaUndo = (sender, e) => { if (richTextBoxCodigo.CanUndo) richTextBoxCodigo.Undo(); };

        private void Form1_Load(object sender, EventArgs e)
        {
            // Sin usar expresión lambda
            buttonAbrir.Click += buttonAbrir_Click;
            // Usando expresión lambda
            buttonAbrir.Click += (object o, EventArgs e) => Abrir(); 

            buttonUndo.Click += variosUndo_Click;
            menuUndo.Click += variosUndo_Click;

            //EventHandler lambdaUndo = (sender, e) => { if(richTextBoxCodigo.CanUndo) richTextBoxCodigo.Undo(); };

            // Usando expresión lambda definida previamente
            buttonUndo.Click += lambdaUndo;
            menuUndo.Click += lambdaUndo;

            // Usando expresión lambda no definida previamente
            // No hay porqué definir los tipos de los parámetros
            buttonUndo.Click += (object sender, EventArgs e) => { if (richTextBoxCodigo.CanUndo) richTextBoxCodigo.Undo(); };

            menuUndo.Click += (sender, e) => { if (richTextBoxCodigo.CanUndo) richTextBoxCodigo.Undo(); };

        }

        private void buttonAbrir_Click(object sender, EventArgs e)
        {
            Abrir();
        }

        private void variosUndo_Click(object sender, EventArgs e)
        {
            if (richTextBoxCodigo.CanUndo) richTextBoxCodigo.Undo();
        }

        private void Abrir()
        {
        }
    }
}

File: Form1.vb
Imports System
Imports System.Windows.Forms

Public Class Form1 : Inherits Form

    Shared Sub Main()
    End Sub

    Private WithEvents buttonAbrir As New Button()
    Private WithEvents richTextBoxCodigo As New RichTextBox()
    Private WithEvents buttonUndo As New Button()
    Private WithEvents menuUndo As New ToolStripMenuItem()

    Private lambdaUndo As EventHandler =
                Sub(sender As Object, e As EventArgs) _
                    If richTextBoxCodigo.CanUndo Then richTextBoxCodigo.Undo()

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles Me.Load
        ' Sin usar expresión lambda
        AddHandler buttonAbrir.Click, AddressOf buttonAbrir_Click
        ' Usando expresión lambda
        AddHandler buttonAbrir.Click, Sub() Abrir()

        ' Usando expresión lambda definida previamente
        AddHandler buttonUndo.Click, lambdaUndo
        AddHandler menuUndo.Click, lambdaUndo

        ' Usando expresión lambda no definida previamente
        AddHandler buttonUndo.Click,
                     Sub() If richTextBoxCodigo.CanUndo Then richTextBoxCodigo.Undo()
        AddHandler menuUndo.Click,
                     Sub() If richTextBoxCodigo.CanUndo Then richTextBoxCodigo.Undo()

    End Sub

    Private Sub buttonAbrir_Click(sender As Object, e As EventArgs) Handles buttonAbrir.Click
        Abrir()
    End Sub

    Private Sub variosUndo_Click(sender As Object, e As EventArgs) _
                       Handles buttonUndo.Click, menuUndo.Click
        If richTextBoxCodigo.CanUndo Then richTextBoxCodigo.Undo()
    End Sub

    Private Sub Abrir()
    End Sub

End Class